[WIP] NFL tutorial draft (#187)
timescale/docs: timescaledb/tutorials/nfl-fantasy-league.md

# Win your NFL fantasy league with TimescaleDB

This tutorial is a step-by-step guide on how to ingest and analyze american football data with TimescaleDB.

> The dataset that we're using is provided be the National Football League (NFL) and contains data about 

Reviewer: "The dataset we're using is provided **BY** the..."

Author: https://github.com/timescale/docs/pull/187/commits/6a12d2bc2512699e3004e5cd0a3da6bf3ce7a39f

> all the passing plays of the 2018 NFL season. We're ingesting this dataset and exploring it in fun ways. You will also see how this dataset and using TimescaleDB can improve your NFL fantasy league strategy. If you aren't an NFL fan, don't worry, you will still find this tutorial helpful as it's showcasing how to get started with TimescaleDB and explore a real world dataset with SQL and Python.

Reviewer: The second sentence feels a little out of place, maybe expand on things more? Or cut it??

Reviewer: "You will also see how this dataset, and using TimescaleDB can improve your..." maybe a comma here?

Author: https://github.com/timescale/docs/pull/187/commits/6a12d2bc2512699e3004e5cd0a3da6bf3ce7a39f

> 1. [Create tables][]

Reviewer: should we add prerequisites and about the data set in the list? It feels a little confusing to have this list then have to wait to see it manifest.

Author: https://github.com/timescale/docs/pull/187/commits/25196ee59ae89b060d067d45367bbbe51b4aaa90

2. [Ingest data from CSV files][] 
3. [Analyze NFL data][]
   
## Prerequisites

* Python 3
* TimescaleDB (see [installation options][install-timescale]) 
* [Psql][psql-install] or any other PostgreSQL client (e.g. DBeaver)
* Pandas and Matplotlib (`pip install pandas matplotlib`)

> ## About the dataset

Reviewer: This section seems a little redundant, it seems we are paraphrasing from above, maybe we should cut it or try to provide new information??

The dataset we're using in this tutorial is provided by the National Football League (NFL) as part of their yearly

> Big Data Bowl. Big Data Bowl is a contest for data professionals and students to explore american football data and build interesting machine learning models.

Reviewer: "THE Big Data Bowl...."?

    birthDate DATE,
    collegeName TEXT,
    position TEXT,
    displayName TEXT
);

CREATE TABLE play (
    gameId INT,
    playId INT,
    playDescription TEXT,
    quarter INT,
    down INT,
    yardsToGo INT,
    possessionTeam TEXT,
    playType TEXT,
    yardlineSide TEXT,
    yardlineNumber INT,
    offenseFormation TEXT,
    personnelO TEXT,
    defendersInTheBox INT,
    numberOfPassRushers INT,
    personnelD TEXT,
    typeDropback TEXT,
    preSnapVisitorScore INT,
    preSnapHomeScore INT,
    gameClock TIME,
    absoluteYardlineNumber INT,
    penaltyCodes TEXT,
    penaltyJerseyNumber TEXT,
    passResult TEXT,
    offensePlayResult INT,
    playResult INT,
    epa DOUBLE PRECISION,
    isDefensivePI BOOLEAN
);

CREATE TABLE tracking (
    time TIMESTAMP,
    x DOUBLE PRECISION,
    y DOUBLE PRECISION,
    s DOUBLE PRECISION,
    a DOUBLE PRECISION,
    dis DOUBLE PRECISION,
    o DOUBLE PRECISION,
    dir DOUBLE PRECISION,
    event TEXT,
    nflId INT,
    displayName TEXT,
    jerseyNumber INT,
    position TEXT,
    frameId INT,
    team TEXT,
    gameId INT,
    playId INT,
    playDirection VARCHAR(5),
    route TEXT
);
```

## Ingest data from CSV files

> We have 3 separate CSV files for game, player, and play tables. For the tracking table, you will need to

Reviewer: "We have three separate CSV files for the game, player, and play tables." maybe?

import data from 17 CSV files (1 file for each week in the season).

Let's use a Python script that utilizes psycopg2's `copy_expert` function to ingest the data:

```python
import config
import psycopg2

> 

Reviewer: should we add code comments to walk users through what each step is doing??
ex: 
"# First connect to the database location"
"# Create function to define cursor and import data"

conn = psycopg2.connect(database=config.DB_NAME, 
                        host=config.HOST, 
                        user=config.USER, 
                        password=config.PASS, 
                        port=config.PORT)

def insert(csv_file, table_name):
    cur = conn.cursor()
    copy_sql = """
            COPY {table} FROM stdin WITH CSV HEADER
            DELIMITER as ','
            """.format(table=table_name)
    with open(csv_file, 'r') as f:
        cur.copy_expert(sql=copy_sql, file=f)
        conn.commit()
        cur.close()

print("Inserting games.csv")
insert("data/games.csv", "game")

print("Inserting plays.csv")
insert("data/plays.csv", "play")

print("Inserting players.csv")
insert("data/players.csv", "player")

for i in range(1, 18): 
    print("Inserting week{i}".format(str(i)))
    insert("data/week{i}.csv".format(i=i), "tracking")

conn.close()
```

## Analyze NFL data

> Now that you have all the data ingested, let's see some ideas how you can analyze the data to help you perfect

Reviewer: "let's go over some ideas on how you can analyze the data to help..."

# Win your NFL fantasy league with TimescaleDB

> This tutorial is a step-by-step guide on how to ingest and analyze american football data with TimescaleDB.

Reviewer 2: ```suggestion
This tutorial is a step-by-step guide on how to ingest and analyze American fantasy football data with TimescaleDB.
```

The dataset that we're using is provided by the National Football League (NFL) and contains data about 
all the passing plays of the 2018 NFL season. We're going to ingest this dataset with Python into TimescaleDB 
and start exploring it to discover interesting things about players that could help you win your next fantasy season. 

> If you aren't an NFL fan, don't worry, you will still find this tutorial helpful as it's showcasing how to 

Reviewer 2: ```suggestion
If you aren't an NFL fan, this tutorial can still help you  
```

get started with TimescaleDB and explore a real world dataset with SQL and Python.

1. [Create tables][]
2. [Ingest data from CSV files][] 
3. [Analyze NFL data][]
   
## Prerequisites

* Python 3
* TimescaleDB (see [installation options][install-timescale]) 
* [Psql][psql-install] or any other PostgreSQL client (e.g. DBeaver)
* Pandas and Matplotlib (`pip install pandas matplotlib`)

## Download the dataset

> Download dataset here!

Reviewer 2: link?

## Create tables

> You will need to create four tables:

Reviewer 2: ```suggestion
You need to create four tables:
```

* game
  
  Information about each game.
* player
  
  Player information.
* play
  
  Play information
* tracking
  

>   Tracking information from each play.

Reviewer 3: Would be great to add a few key fields from each table or something to describe this data just a little bit more.

```sql
CREATE TABLE game (
    gameId INT PRIMARY KEY,
    gameDate DATE,
    gameTimeEastern TIME,
    homeTeamAbbr VARCHAR(3),

>     visitorTeamAbbr VARCHAR(3),

Reviewer 3: Guaranteed to always bee 3 characters? Again, in general, the Postgres world uses TEXT for almost all text unless there is a programatic needs to have the DB limit/check the size of data and reject.

    height TEXT,
    weight INT,
    birthDate DATE,
    collegeName TEXT,
    position TEXT,
    displayName TEXT
);

CREATE TABLE play (
    gameId INT,
    playId INT,
    playDescription TEXT,
    quarter INT,
    down INT,
    yardsToGo INT,
    possessionTeam TEXT,
    playType TEXT,
    yardlineSide TEXT,
    yardlineNumber INT,
    offenseFormation TEXT,
    personnelO TEXT,
    defendersInTheBox INT,
    numberOfPassRushers INT,
    personnelD TEXT,
    typeDropback TEXT,
    preSnapVisitorScore INT,
    preSnapHomeScore INT,
    gameClock TIME,
    absoluteYardlineNumber INT,
    penaltyCodes TEXT,
    penaltyJerseyNumber TEXT,
    passResult TEXT,
    offensePlayResult INT,
    playResult INT,
    epa DOUBLE PRECISION,
    isDefensivePI BOOLEAN
);

CREATE TABLE tracking (
    time TIMESTAMP,
    x DOUBLE PRECISION,
    y DOUBLE PRECISION,
    s DOUBLE PRECISION,
    a DOUBLE PRECISION,
    dis DOUBLE PRECISION,
    o DOUBLE PRECISION,
    dir DOUBLE PRECISION,
    event TEXT,
    nflId INT,
    displayName TEXT,
    jerseyNumber INT,
    position TEXT,
    frameId INT,
    team TEXT,
    gameId INT,
    playId INT,
    playDirection VARCHAR(5),
    route TEXT
);
```

> Add indexes to the `tracking` table speed up queries:

Reviewer 3: ```suggestion
Add indexes to the `tracking` table to improve query performance:
```

```sql
CREATE INDEX idx_gameid ON tracking (gameid);
CREATE INDEX idx_playerid ON tracking (player_id);
CREATE INDEX idx_playid ON tracking (playid);
```

## Ingest data from CSV files

> We have three separate CSV files for game, player, and play tables. For the tracking table, you will need to

Reviewer 2: ```suggestion
There are three separate CSV files for game, player, and play tables. For the tracking table, you need to
```

> import data from 17 CSV files (1 file for each week in the season).

Reviewer 2: ```suggestion
import data from 17 CSV files (one file for each week of the season).
```

> Let's use a Python script that utilizes psycopg2's `copy_expert` function to ingest the data:

Reviewer 2: ```suggestion
You can use a Python script that uses psycopg2's `copy_expert` function to ingest the data:
```

Reviewer 3: So, dumb question here. Is it worth showing an example with `COPY` or psql `/copy`?

```python
import config
import psycopg2

# connect to the database
conn = psycopg2.connect(database=config.DB_NAME, 
                        host=config.HOST, 
                        user=config.USER, 
                        password=config.PASS, 
                        port=config.PORT)

# insert CSV file into given table
def insert(csv_file, table_name):
    cur = conn.cursor()
    copy_sql = """
            COPY {table} FROM stdin WITH CSV HEADER
            DELIMITER as ','
            """.format(table=table_name)
    with open(csv_file, 'r') as f:
        cur.copy_expert(sql=copy_sql, file=f)
        conn.commit()
        cur.close()

print("Inserting games.csv")
insert("data/games.csv", "game")

print("Inserting plays.csv")
insert("data/plays.csv", "play")

print("Inserting players.csv")
insert("data/players.csv", "player")

# iterate over each week's CSV file and insert
for i in range(1, 18): 
    print("Inserting week{i}".format(str(i)))
    insert("data/week{i}.csv".format(i=i), "tracking")

conn.close()
```

## Analyze NFL data

> Now that you have all the data ingested, let's go over some ideas on how you can analyze the data to help you perfect

Reviewer 2: ```suggestion
Now that you have all the data ingested, here are some ideas for analyzing the data to help you perfect
```

your fantasy drafting strategy and win your fantasy season.

> To analyze the dataset you will need to install two libraries, if you haven't already: Pandas and Matplotlib.

Reviewer 2: ```suggestion
To analyze the dataset you need to install two libraries, if you haven't already: Pandas and Matplotlib.
```

Reviewer 3: This sounds like a requirement, which it shouldn't be (generally). Feels like we should say something like, "For each example query below, we visualize the result with Matplotlib in Python. If you want to graph the data as well, you will need to import Pandas and Matplotlib."

Remember, we want to emphasize the SQL part of this, but so far I feel like I'm only reading a Python tutorial. To me, I think it needs to be clear up front that we're analyzing the data with SQL, and using Python as a way to interact with/visualize the results. IMO

### Install pandas and matplotlib
```bash
pip install pandas matplotlib
```

1. Number of yards run in game for passing plays, by player and game 

> 1. Avg yards run for a player over a game

Reviewer 2: ```suggestion
1. Average yards run for a player over a game
```

> 1. Average and median yards run per game by type of player (not taking avg of individual)

Reviewer 2: ```suggestion
1. Average and median yards run per game by type of player (not taking the average of an individual)
```

1. Num of snap plays by player where they were on the offense
1. Number of plays vs points scored

> 1. Average yards per game for top 3 players of each position

Reviewer 2: ```suggestion
1. Average yards per game for top three players of each position
```
